crates/oxc_linter/src/rules/typescript/no_var_requires.rs

```diff
@@ -0,0 +1,115 @@
+)]
+#[diagnostic(severity(error))]
+struct NoVarRequiresDiagnostic(#[label] pub Span);
+
+#[derive(Debug, Default, Clone)]
+pub struct NoVarRequires;
+
+declare_oxc_lint!(
+    /// ### What it does
+    /// 
+    /// Disallow `require` statements except in import statements
+    /// 
+    /// ### Why is this bad?
+    /// 
+    /// In other words, the use of forms such as var foo = require("foo") are banned. Instead use ES6 style imports or import foo = require("foo") imports.
+    /// 
+    /// ```javascript
+    /// var foo = require('foo');
+    /// const foo = require('foo');
+    /// let foo = require('foo');
+    /// ```
+    NoVarRequires,
+    correctness
+);
+
+impl Rule for NoVarRequires {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        if let AstKind::CallExpression(expr) = node.kind() && expr.is_require_call() {
+            if ctx.scopes().get_bindings(node.scope_id()).contains_key("require") {
+                return;
+            } 
+
+            if let Some(parent_node) = ctx.nodes().parent_node(node.id()) {
+                match parent_node.kind() {
+                    AstKind::Argument(_) => {
+                        if let Some(parent_node) = ctx.nodes().parent_node(parent_node.id()) {
+                            if is_target_node(&parent_node.kind()) {
+                                ctx.diagnostic(NoVarRequiresDiagnostic(expr.span));
+                            }
+                        }
+                    }
+                    _ => {}
+                }
+
+                if is_target_node(&parent_node.kind()) {
+                    ctx.diagnostic(NoVarRequiresDiagnostic(expr.span));
+                }
+            }
+        }
+    }
+}
+
+fn is_target_node<'a>(node_kind: &AstKind<'a>) -> bool {
+    match node_kind {
+        AstKind::CallExpression(_)
+        | AstKind::MemberExpression(_)
+        | AstKind::NewExpression(_)
+        | AstKind::TSAsExpression(_)
+        | AstKind::TSTypeAssertion(_)
+        | AstKind::VariableDeclarator(_) => {
+            return true;
+        }
+        _ => return false
+    }
+}
+
+#[test]
+fn test() {
+    use crate::tester::Tester;
+
+    let pass = vec![
+        "import foo = require('foo');",
+        "require('foo');",
+        "require?.('foo');",
+        r#"
+            import { createRequire } from 'module';
+            const require = createRequire('foo');
+            const json = require('./some.json');
+        "#,
+    ];
+
+    let fail = vec![
+        "var foo = require('foo');",
+        "const foo = require('foo');",
+        "let foo = require('foo');",
+        "let foo = trick(require('foo'));",
+        "var foo = require?.('foo');",
+        "const foo = require?.('foo');",
+        "let foo = require?.('foo');",
+        "let foo = trick(require?.('foo'));",
+        "let foo = trick?.(require('foo'));",
+        "const foo = require('./foo.json') as Foo;",
+        // "const foo = <Foo>require('./foo.json');",
```

Comment:
I commented it out for the time being because it would result in a syntax error.
Should I raise an issue 🤔 ?

https://boshen.github.io/oxc/playground/?code=dAB5AHAAZQAgAEYAbwBvACAAPQAgAHsAfQAKAGMAbwBuAHMAdAAgAGYAbwBvACAAPQAgADwARgBvAG8APgByAGUAcQB1AGkAcgBlACgAJwAuAC8AZgBvAG8ALgBqAHMAbwBuACcAKQA7AAoA

Reply:
https://www.typescriptlang.org/docs/handbook/jsx.html#the-as-operator

We parsing everything as `.tsx`, you can add a comment to justify this commented code.

Reply:
😮  I see. Thanks so much. I have added a comment.